generic-token-tests/tests/test_generic_token.rs

```diff
@@ -0,0 +1,236 @@
+use {
+    rand::prelude::*,
+    spl_generic_token::{generic_token, token, token_2022},
+    spl_token::{
+        solana_program::program_pack::Pack,
+        state::{Account as SplAccount, AccountState as SplAccountState, Mint as SplMint},
+    },
+    spl_token_2022::{
+        extension::set_account_type,
+        state::{Account as SplAccount2022, Mint as SplMint2022, Multisig as SplMultisig},
+    },
+    test_case::test_case,
+};
+
+fn random_token_account() -> SplAccount {
+    let mut rng = thread_rng();
+
+    let mint = solana_pubkey::new_rand();
+    let owner = solana_pubkey::new_rand();
+    let amount = rng.gen();
+    let delegate = if rng.gen() {
+        Some(solana_pubkey::new_rand())
+    } else {
+        None
+    }
+    .into();
+    let state = rng.gen_range(0..3).try_into().unwrap();
+    let is_native = rng.gen::<Option<u64>>().into();
+    let delegated_amount = rng.gen();
+    let close_authority = if rng.gen() {
+        Some(solana_pubkey::new_rand())
+    } else {
+        None
+    }
+    .into();
+
+    SplAccount {
+        mint,
+        owner,
+        amount,
+        delegate,
+        state,
+        is_native,
+        delegated_amount,
+        close_authority,
+    }
+}
+
+#[test_case(false; "spl_token")]
+#[test_case(true; "spl_token_2022")]
+fn test_generic_account(is_token_2022_account: bool) {
+    for _ in 0..1000 {
+        let expected_account = random_token_account();
+        let is_initialized = expected_account.state != SplAccountState::Uninitialized;
+
+        let mut account_data = vec![0; SplAccount::LEN];
+        expected_account.pack_into_slice(&mut account_data);
+
+        // check the basic rules of the parser:
+        // * uninitialized accounts never parse
+        // * standard token accounts parse as both
+        // * typed 2022 accounts parse only as 2022
+        if is_initialized && is_token_2022_account {
+            account_data.resize(SplAccount::LEN + 2, 0);
+            set_account_type::<SplAccount2022>(&mut account_data).unwrap();
```

Comment:
to be honest i couldnt figure out why this expects to see two extra bytes instead of one.

Reply:
I think that's actually an off-by-one error in `type_and_tlv_indices`, and the `<=` should be `<`: https://github.com/solana-program/token-2022/blob/f2341f9d06abc0f0781cec5948232169d5e372a2/program/src/extension/mod.rs#L315

I put in https://github.com/solana-program/token-2022/pull/369 to fix it